UBL: G29 P2 & P4 bug fixes & improvements

Bob-the-Kuhn opened the pull request:
**P2 changes**
- Business card mode - set SIZE_OF_LITTLE_RAISE to 1 so user can get the shim out from under the nozzle
- changed to UBL_MESH_MAX/MIN in a couple of places so that the warnings applied to the bed boundries, not the travel limits
-added a P2 finished serial print to the host so I stop wondering why it's no longer responding to the encoder wheel.
- added check to beginning of "use_encoder_wheel_to_measure_point()" so that a long press or a bounce won't end this routine before any work is done.  This is only needed the second time through because there's no stepper movement between the encoder click and invoking this routine the second time.  Usually the stepper movements are long enough that the bounce is over with before entering this type of routine
- "measure_business_card_thickness" moved printed instructions to after the stepper movements have finished & added stepper.synchronize
- "measure_business_card_thickness" changed the location from the middle of the travel area to the middle of the bed
- "manually_probe_remaining_mesh"  changed "probe off bed check from travel limits to bed limits
- "manually_probe_remaining_mesh"  added check for user holding down the encoder button too long just before doing the measurement (the check for holding down the encoder button for a LONG time is done after the measurement)

**P4**
- added check for unprobed point (same as in PR #6357)
- moved stepper commands to after the unprobed point check so the nozzle doesn't go to the unprobed points

Bob-the-Kuhn committed:
G29 P2 & P4 bug fixes & improvements

=========================================================

formatting clean up & eliminated X_MAX_POS, ...

Bob-the-Kuhn commented:
Question on the P4 command.  The documentation says the nozzle will move when you spin the encoder wheel but that doesn't happen.

Is the documentation out of date or do we need to fix the code?

Roxy-3D commented:
> Question on the P4 command. The documentation says the nozzle will move when you spin the encoder wheel but that doesn't happen.  Is the documentation out of date or do we need to fix the code?

The documentation is out of date.   But a number of people have asked for this.   I originally had this in the code.  But I took it out for several reasons.   First...  The machine I was doing development on has a super slow Z-Axis and it just slowed me down trying to get the mesh tuned.   But the other reason is it complicates things because if the person lowers the nozzle, what do you do?  Do you raise it before moving to the next spot?  I think you probably have to.

But all that aside...   We probably should add an option to 
```cpp
#define UBL_MOVE_NOZZLE_DURING_MESH_EDIT
```

thinkyhead commented:
Lots of other fixes needed. Indentation, application of coding standards, use of helpful macros like `sq()` need to be applied, etc. Will follow up with further cleanups after merging this.

thinkyhead commented on Marlin/ubl_G29.cpp:
```diff
@@ -455,8 +457,9 @@
-        } break;
+          SERIAL_PROTOCOLLNPGM("G29 P2 finished");
+          }
+          break;
```
Indentation is screwed up here.

thinkyhead commented on Marlin/ubl_G29.cpp:
```diff
@@ -917,15 +924,18 @@
-    SERIAL_PROTOCOLLNPGM("Place Shim Under Nozzle and Perform Measurement.");
     do_blocking_move_to_z(in_height);
-    do_blocking_move_to_xy((float(X_MAX_POS) - float(X_MIN_POS)) / 2.0, (float(Y_MAX_POS) - float(Y_MIN_POS)) / 2.0);
+    do_blocking_move_to_xy((float(UBL_MESH_MAX_X) - float(UBL_MESH_MIN_X)) / 2.0, (float(UBL_MESH_MAX_Y) - float(UBL_MESH_MIN_Y)) / 2.0);
```
Slightly more concise:
```cpp
do_blocking_move_to_xy(float(UBL_MESH_MAX_X - (UBL_MESH_MIN_X)) / 2.0, float(UBL_MESH_MAX_Y - (UBL_MESH_MIN_Y)) / 2.0);
```
This also accomplishes the same with no casting needed:
```cpp
do_blocking_move_to_xy(0.5 * (UBL_MESH_MAX_X - (UBL_MESH_MIN_X)), 0.5 * (UBL_MESH_MAX_Y - (UBL_MESH_MIN_Y)));
```

thinkyhead commented on Marlin/ubl_G29.cpp:
```diff
@@ -984,6 +996,10 @@
       if (do_ubl_mesh_map) ubl.display_map(map_type);  // show user where we're probing
 
+
+      while (ubl_lcd_clicked()) delay(50);;  // wait for user to release encoder wheel
```
It looks like your keyboard is malfunctioning. Doubled-up returns, spaces, and other characters keep getting inserted.

thinkyhead commented on Marlin/ubl_G29.cpp:
```diff
@@ -1436,33 +1454,52 @@
+
+
+
+        ubl.has_control_of_lcd_panel = true; // There is a race condition for the Encoder Wheel getting clicked.
```
This amount of blank lines is excessive.

thinkyhead replied:
After the user-wait is over, return to normal state with KEEPALIVE_STATE(IN_HANDLER).

thinkyhead commented on Marlin/ubl_G29.cpp:
```diff
@@ -1436,33 +1454,52 @@
+
+        ubl.has_control_of_lcd_panel = true; // There is a race condition for the Encoder Wheel getting clicked.
+                                             // It could get detected in lcd_mesh_edit (actually _lcd_mesh_fine_tune)
+                                             // or here.
```
If end-of-line comments need to span several lines, just make them into normal comments in front of the code. It's easier to read.

thinkyhead commented on Marlin/ubl_G29.cpp:
```diff
@@ -900,6 +903,10 @@
 
   float use_encoder_wheel_to_measure_point() {
+
+    while (ubl_lcd_clicked()) delay(50);;  // wait for user to release encoder wheel
```
Extra semicolon.

thinkyhead commented on Marlin/ubl_G29.cpp:
```diff
@@ -1436,33 +1454,52 @@
+        do {
+          new_z = lcd_mesh_edit();
+          idle();
+        } while (!ubl_lcd_clicked());
```
After the user-wait is over, return to normal state with `KEEPALIVE_STATE(IN_HANDLER)`.

thinkyhead commented on Marlin/ubl_G29.cpp:
```diff
@@ -1436,33 +1454,52 @@
+
+
+
+
```
This number of blank lines is excessive.

thinkyhead commented:
Rebased and augmented at #6552

Bob-the-Kuhn commented:
I believe I've implemented all the comments except the one about `KEEPALIVE_STATE(IN_HANDLER)`.  I couldn't find where it should go.

I've re-indented the file, changed all multiple blank lines to a single blank line and all double semi-colons to single.

I've also replaces all the  X_MIN_POS ... travel limits with the UBL_MESH_MIN_X ... versions

I wasn't able to find any place that needs the sq() function.  I did see that it was used in the file.

I believe that the extra blank lines came about because I saved the file while still in the Notepad++ compare mode.

Bob-the-Kuhn commented:
Just can't keep up with @thinkyhead

thinkyhead commented:
Pushing through these last PRs quickly, so I can write up release notes, apply tags, etc. ahead of release, during which others will have a chance to review changes.
